### ISSUE TITLE
Getting warning bad fde: FDE is really a CIE - GCC 6.3.1 extern "C"

### ISSUE THREAD

**Issue by suedama1756** [llvm/llvm-project#86]
Environment:

* Redhat Enterprise Server Linux Server release 7.6
* G++: 6.3.1 20170216 (Redhat 6.3.1-3)
* LLVM 9.0

I'm getting an "bad fde: FDE is really a CIE" warning when making calls to extern "C" functions in GCC 6.3.1.  I can make the issue go away if I mark the external function with NoUnwind, however, I don't think this is the correct solution as the function may throw.

Taking a wild guess  I tried setting the personality of the external function to __gxx_personality_v0 however this then fails validation as external functions cannot be given a personality.

I'm using LLVM 9.0 as a JIT, I'm using the ORC2 API, however, the issue is also demonstrable using the Kaleidascope JIT here: https://llvm.org/docs/tutorial/index.html The error is raised from the DwarfParser.hpp at the point the generated function is resolved (jitted)

Is there a correct way to deal with this, is it something I need to even worry about?

**Comment by dwblaikie** [comment]
you don't mention how any of this relates to clang/llvm - where is the error being produced from? where is llvm/clang used in this experiment?

**Comment by suedama1756** [comment]
> you don't mention how any of this relates to clang/llvm - where is the error being produced from? where is llvm/clang used in this experiment?

I kind of figured it would be obvious to those that new LLVM, however, in hindsight a little more information was probably required. Question is updated...

**Comment by lhames** [comment]
At a minimum it’s usually helpful to provide your platform (Linux, Windows, Darwin, etc.), platform version, the LLVM version you are using and a little bit of context on what you were trying to do when you saw the error. Your update to the question helps a lot. Could you also include the platform you are running on?

Are you able to run this under a debugger and obtain a backtrace for the error? It’s coming from libunwind, so if you have the source for your LLVM build you should be able to grep it for “FDE is really a CIE” to find the line number to break on.

If this is happening in the JIT then my initial best guess is a bug in the eh-frame parsing / re-writing logic.

**Comment by suedama1756** [comment]
I've updated with version information. I do not have the full source for LLVM build I'm using, its a corporate box which I have limited access to so may take me a while to sort out. In the mean time do you have any ideas on a) are there any function attributes I can explicitly set other than nounwind that could help here. b) is this an issue I need to even care about?

**Comment by lhames** [comment]
Very belatedly: I think this is a common issue when using libunwind with JIT'd code on Linux. For a long time the JIT used "host-is-linux" as a proxy for "unwinder is libgcc_s". Since libgcc_s's implementation of `__register_frame` provided a fast-path if you registered the whole eh-frame section that's what we did. The problem happens when we call `__register_frame` this way and the unwinder is actually libunwind: The eh-frame section starts with a CIE, and it triggers this error.

@suedama1756 I think it is a problem -- we won't try to register the subsequent frames, so exceptions thrown through the JIT'd code will not behave correctly.

We're in the process of working through a fix for this: bab39816085 added new functions to libunwind that have the behavior we want. We need to update the JIT to detect when these functions are available and use them. This will allow the JIT to work correctly with newer versions of libunwind. (Versions before bab39816085 probably won't be fixed).

**Comment by lygstate** [comment]
> Very belatedly: I think this is a common issue when using libunwind with JIT'd code on Linux. For a long time the JIT used "host-is-linux" as a proxy for "unwinder is libgcc_s". Since libgcc_s's implementation of `__register_frame` provided a fast-path if you registered the whole eh-frame section that's what we did. The problem happens when we call `__register_frame` this way and the unwinder is actually libunwind: The eh-frame section starts with a CIE, and it triggers this error.
> 
> @suedama1756 I think it is a problem -- we won't try to register the subsequent frames, so exceptions thrown through the JIT'd code will not behave correctly.
> 
> We're in the process of working through a fix for this: [bab3981](https://github.com/llvm/llvm-project/commit/bab39816085d715e52c2131fa249ccd10178764b) added new functions to libunwind that have the behavior we want. We need to update the JIT to detect when these functions are available and use them. This will allow the JIT to work correctly with newer versions of libunwind. (Versions before [bab3981](https://github.com/llvm/llvm-project/commit/bab39816085d715e52c2131fa249ccd10178764b) probably won't be fixed).

I think I also faced this issue, I am using mesa/llvmpipe that used llvm jit, and it's raised error:
```
libunwind: __unw_add_dynamic_fde: bad fde: FDE is really a CIE
libunwind: __unw_add_dynamic_fde: bad fde: FDE is really a CIE
libunwind: __unw_add_dynamic_fde: bad fde: FDE is really a CIE
libunwind: __unw_add_dynamic_fde: bad fde: FDE is really a CIE
libunwind: __unw_add_dynamic_fde: bad fde: FDE is really a CIE
libunwind: __unw_add_dynamic_fde: bad fde: FDE is really a CIE
```
What should  I need  to do to fixes this? #55584

**Comment by lhames** [comment]
@lygstate -- @housel committed fixes for libunwind (bab39816085d7), and LLVM and the ORC runtime (981523b2e4675d4be496a333583b2f728c057d02) that address this issue in the case where exceptions are registered via the ORC runtime.

You could use the ORC runtime in your JIT to fix this issue. We should also add

**Comment by lhames** [comment]
I've filed https://github.com/llvm/llvm-project/issues/55681 to teach LLVM to use @housel's registration functions.

**Comment by DoDoENT** [comment]
Are there any workarounds or at least a way to silence this?

I've compiled neovim with a custom-built LLVM toolchain (15.0.6) and its luaJIT engine keeps popping this warning all over the place. It makes the editor unusable...

**Comment by DoDoENT** [comment]
Interestingly, libunwind seems to log that using [`_LIBUNWIND_DEBUG_LOG`](https://github.com/llvm/llvm-project/blob/main/libunwind/src/libunwind.cpp#L311) macro which should be [no-op in non-debug builds](https://github.com/llvm/llvm-project/blob/main/libunwind/src/config.h#L183-L218), and I definitely compiled my toolchain in [release mode](https://github.com/microblink/centos-clang/blob/master/Dockerfile#L81-L118).

It appears some cmake script somewhere is naughty and forcibly removes `NDEBUG` in release builds...

**Comment by DoDoENT** [comment]
Ah, assertions are always [enabled for libunwind](https://github.com/llvm/llvm-project/blob/main/libunwind/CMakeLists.txt#L37). I'll disable them in my build until this issue gets resolved...

**Comment by lhames** [comment]
@mortezadadgar Are you running on Darwin or Linux? If Linux, did you mean to use libunwind as the unwinder?

**Comment by mortezadadgar** [comment]
I use gentoo just switched over clang profile I suppose this profile uses libunwind

```
$ portageq envvar LDFLAGS
-Wl,-O1 -Wl,--as-needed -fuse-ld=lld -rtlib=compiler-rt -unwindlib=libunwind -Wl,--as-needed
```

**Comment by DoDoENT** [comment]
> Have you found any solution i'm facing same issue

Didn't bother. I'm using my own custom toolchain that I built without assertions so it's silent. And apparently everything works, so it may be that this is just a cosmetic warning...

**Comment by mortezadadgar** [comment]
@DoDoENT which option disables assertions

**Comment by DoDoENT** [comment]
@mortezadadgar `-DLIBUNWIND_ENABLE_ASSERTIONS=OFF`. See my [dockerfile](https://github.com/microblink/centos-clang/blob/467f119949bf9ce834857d7ff1ca5b24296df2c7/Dockerfile#L119).

**Comment by mortezadadgar** [comment]
> @mortezadadgar `-DLIBUNWIND_ENABLE_ASSERTIONS=OFF`. See my [dockerfile](https://github.com/microblink/centos-clang/blob/467f119949bf9ce834857d7ff1ca5b24296df2c7/Dockerfile#L119).

that's already is disabled in llvm-libunwind ebuild, I suppose it's a bug in clang-15.0.7

**Comment by soleera** [comment]
> > @mortezadadgar `-DLIBUNWIND_ENABLE_ASSERTIONS=OFF`. See my [dockerfile](https://github.com/microblink/centos-clang/blob/467f119949bf9ce834857d7ff1ca5b24296df2c7/Dockerfile#L119).
> 
> that's already is disabled in llvm-libunwind ebuild, I suppose it's a bug in clang-15.0.7

**TL;DR: This is a Gentoo-specific bug which is unrelated to the bug in libunwind that @lhames was talking about!**

This is due to the cmake eclass [clearing all flags](https://github.com/gentoo/gentoo/blob/f042f7c1d9a26422514494a1d4e98e51a4255b28/eclass/cmake.eclass#L559) from the selected build type ([defaults to RelWithDebInfo in EAPI 8](https://github.com/gentoo/gentoo/blob/f042f7c1d9a26422514494a1d4e98e51a4255b28/eclass/cmake.eclass#L53)). The intent was to remove the default compiler optimizations, but unfortunately this is also where `-DNDEBUG` is defined. At the same time, the conditional logic in libunwind's [CMakeLists.txt](https://github.com/llvm/llvm-project/blob/9e37142dc12a7d57a793db56ea5fd259171fe20a/libunwind/CMakeLists.txt#L237) relies on the assumption that NDEBUG will only be undefined in DEBUG build types; when this assumption is broken, `LIBUNWIND_ENABLE_ASSERTIONS` does nothing.

This same issue has occurred with [other LLVM ebuilds](https://bugs.gentoo.org/614844), and there [was a suggestion](https://bugs.gentoo.org/884791) to re-enable `-DNDEBUG` in cmake.ebuild by default, however the[ current position](https://bugs.gentoo.org/884791#c8) seems to be that this is something to be done in ebuilds on a per-package basis. As such this should be considered a bug with the llvm-libunwind ebuild on Gentoo; in the meantime, all you need to do is copy the llvm-libunwind ebuild to a local repository, and add the following code under `multilib_src_configure()`:

```
	# Respect upstream build type assumptions
	# https://github.com/llvm/llvm-project/issues/86
	# Related to #614844 and #884791
	[[ "${CMAKE_BUILD_TYPE^^}" == DEBUG ]] || append-flags -DNDEBUG
```

**Comment by thesamesam** [comment]
Fixed on the Gentoo side in https://github.com/gentoo/gentoo/commit/f8fa80bed10eacdc4269f6d8a08985a6cf6df1dc, thanks!

**Comment by theoparis** [comment]
Still running into this with my own custom LLVM 19 (HEAD) build... `-DLIBUNWIND_ENABLE_ASSERTIONS=OFF` and `-DCMAKE_C_FLAGS="-DNDEBUG"` + `-DCMAKE_CXX_FLAGS="-DNDEBUG" didn't seem to fix neovim.
